Save for later: Add 'remove from saved posts' animation

ctarda opened the pull request:
Implements #9474 

An early PR to validate the general direction.
![animation_recording mov](https://user-images.githubusercontent.com/2722505/40821390-7cf1a2ca-651b-11e8-9105-606973787146.gif)

Implement removal animation and undo by not actually removing the flag marking `ReaderPost`s a saved for later, but by tracking the post that have been selected for removal, and presenting a cell containing their "collapsed state" (title, and undo button) in their place.

Caveats: 
- The design of the collapsed cell is not final. I run out of time and I wanted to validate the rest of the implementation as soon as possible.

To test:
1. Mark a few posts as saved for later
2. Navigate to Saved Posts
3. Tap the "Save for Later" button (bookmark icon) in a few posts. The posts should collapse.
4. Tap Undo. The post should be visible again
5. Navigate away from Saved Posts. Posts that were "removed" should not be visible anymore when navigating back to Saved Posts.

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderSavedPostCellActions.swift:
```diff
@@ -0,0 +1,35 @@
+        removedPosts.remove(post)
+    }
+
+    func clear() {
```
I wonder if this name could be a little more descriptive. `clearRemovedPosts`?

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderSavedPostsViewController.swift:
```diff
@@ -50,6 +57,11 @@ final class ReaderSavedPostsViewController: UITableViewController {
 
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        postCellActions?.clear()
```
I'm not sure that `viewWillDisappear` is the right place to clear the actions. This will currently clear the actions on any navigation away from this screen – whether that's tapping on a different tab bar tab, or tapping into the detail of a saved post.

On Android, I noticed that they only remove the 'undo' cells if you navigate _back_ out of the Saved Posts list. If you tap into the detail of a post and then back to the Saved Posts list, the undo items are still there.

Would `deinit` work? Or `viewDidDisappear`, but check that we're no longer the top view controller in the navigation stack?

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderSavedPostCellActions.swift:
```diff
@@ -0,0 +1,35 @@
+        return removedPosts.contains(post)
+    }
+
+    func remove(_ post: ReaderPost) {
```
How about `restoreUnsavedPost`?

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderSavedPostCellActions.swift:
```diff
@@ -0,0 +1,35 @@
+        delegate?.willRemove(cell)
+    }
+
+    func contains(_ post: ReaderPost) -> Bool {
```
How about `postWasRemoved` or `postIsRemoved`?

frosty commented:
Oh, just noticed there are a couple of horizontal constraint issues in the undo cell xib too. We should resolve those.

frosty commented:
Just updated the gist to fix a couple of issues with the margins of the undo cell, and preventing compression of the Undo button.

ctarda commented:
I've just pushed an update that I think addresses all the comments. Thanks for taking care of the cell layout! 👍 

![simulator screen shot - iphone 8 - 2018-06-01 at 18 52 14](https://user-images.githubusercontent.com/2722505/40869281-070ac476-65cd-11e8-9ad3-843f4bfcd5a3.png)

frosty commented:
Hope you don't mind – I pushed a fix! Feel free to revert if you don't like it :)

ctarda commented:
Thanks for fixing the warning!
